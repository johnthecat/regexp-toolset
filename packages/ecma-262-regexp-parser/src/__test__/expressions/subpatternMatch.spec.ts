import { expect, describe, it } from 'vitest';
import { parseRegexp } from '../../index.js';

/**
 * @see parseGroup
 * @see parseSubpatternMatch
 */
describe('Subpattern Match', () => {
  it('should refer to group with same name', () => {
    expect(parseRegexp(/(?<my_name>a) \k<my_name>/)).toMatchSnapshot();
  });

  it('should fallback to escaped char', () => {
    expect(parseRegexp(/\k/)).toMatchSnapshot('no end');
    expect(parseRegexp(/\k</)).toMatchSnapshot('with opened chevron');
    expect(parseRegexp(/\k<>/)).toMatchSnapshot('with opened and closed chevrons');
    expect(parseRegexp(/\k<hello/)).toMatchSnapshot('with opened chevron and word');
  });

  it('should throw if there is no group with such name', () => {
    expect(() => parseRegexp('/\\k<my_name>/')).toThrowErrorMatchingInlineSnapshot(`
      "
       ❱ /\\\\k<my_name>/
          ═══════════
       This token references a non-existent or invalid subpattern"
    `);
  });
});
