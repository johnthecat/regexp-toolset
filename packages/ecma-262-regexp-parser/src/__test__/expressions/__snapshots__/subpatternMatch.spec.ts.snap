// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Subpattern Match > should fallback to escaped char > no end 1`] = `
{
  "body": {
    "charCode": 107,
    "end": 2,
    "kind": 6,
    "start": 1,
    "type": "escaped",
    "value": "k",
  },
  "end": 3,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Subpattern Match > should fallback to escaped char > with opened and closed chevrons 1`] = `
{
  "body": {
    "end": 4,
    "expressions": [
      {
        "charCode": 107,
        "end": 2,
        "kind": 6,
        "start": 1,
        "type": "escaped",
        "value": "k",
      },
      {
        "charCode": 60,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "<",
      },
      {
        "charCode": 62,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": ">",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 5,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Subpattern Match > should fallback to escaped char > with opened chevron 1`] = `
{
  "body": {
    "end": 3,
    "expressions": [
      {
        "charCode": 107,
        "end": 2,
        "kind": 6,
        "start": 1,
        "type": "escaped",
        "value": "k",
      },
      {
        "charCode": 60,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "<",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 4,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Subpattern Match > should fallback to escaped char > with opened chevron and word 1`] = `
{
  "body": {
    "end": 8,
    "expressions": [
      {
        "charCode": 107,
        "end": 2,
        "kind": 6,
        "start": 1,
        "type": "escaped",
        "value": "k",
      },
      {
        "charCode": 60,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "<",
      },
      {
        "charCode": 104,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": "h",
      },
      {
        "charCode": 101,
        "end": 5,
        "kind": 6,
        "start": 5,
        "type": "simple",
        "value": "e",
      },
      {
        "charCode": 108,
        "end": 6,
        "kind": 6,
        "start": 6,
        "type": "simple",
        "value": "l",
      },
      {
        "charCode": 108,
        "end": 7,
        "kind": 6,
        "start": 7,
        "type": "simple",
        "value": "l",
      },
      {
        "charCode": 111,
        "end": 8,
        "kind": 6,
        "start": 8,
        "type": "simple",
        "value": "o",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 9,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Subpattern Match > should refer to group with same name 1`] = `
{
  "body": {
    "end": 25,
    "expressions": [
      {
        "body": {
          "charCode": 97,
          "end": 12,
          "kind": 6,
          "start": 12,
          "type": "simple",
          "value": "a",
        },
        "end": 13,
        "kind": 23,
        "specifier": {
          "end": 11,
          "kind": 24,
          "name": "my_name",
          "start": 3,
        },
        "start": 1,
        "type": "capturing",
      },
      {
        "charCode": 32,
        "end": 14,
        "kind": 6,
        "start": 14,
        "type": "simple",
        "value": " ",
      },
      {
        "end": 25,
        "groupName": "my_name",
        "kind": 11,
        "ref": {
          "body": {
            "charCode": 97,
            "end": 12,
            "kind": 6,
            "start": 12,
            "type": "simple",
            "value": "a",
          },
          "end": 13,
          "kind": 23,
          "specifier": {
            "end": 11,
            "kind": 24,
            "name": "my_name",
            "start": 3,
          },
          "start": 1,
          "type": "capturing",
        },
        "start": 15,
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 26,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;
